[MRG] ENH: parallel RF oob scoring

tylerjereddy opened the pull request:
* Fixes #28059

* use `joblib` to make oob predictions in random forest

* I also drafted in some `asv` benchmarks to try to capture the performance improvements more formally, but it looks like the `virtualenv` machinery for running `asv` may need a tune up? I kept seeing `Failed: trying different commit/environment` via `asv continuous -E virtualenv -e -b "RandomForestClassifierBenchmark" main treddy_issue_28059` and that was after having to apply a patch to `asv.conf.json` for a missing dependency (perhaps the team more commonly uses `conda` here)

* in lieu of the more formal benchmarks working (it would be good to confirm those...) I re-ran the code from the matching issue and the scaling does look better now with 16 cores on that machine:

<details>

![bench_feat](https://github.com/scikit-learn/scikit-learn/assets/7903078/c5a2fef3-9467-406a-8dec-fc19ac0d6c5b)

</details>

tylerjereddy committed:
[WIP] ENH: parallel RF oob scoring

* Fixes #28059

* use `joblib` to make oob predictions in random forest

adam2392 commented on sklearn/ensemble/_forest.py:
```diff
@@ -599,14 +614,12 @@ def _compute_oob_predictions(self, X, y):
+                delayed(self._get_oob_pred_parallel)(e, n_samples, n_samples_bootstrap, X)
+            for e in self.estimators_)
+
+        for unsampled_indices, y_pred in all_pred:
```
Why not include all the other output below as part of the parallelization? It seems inefficient to have to loop twice over `all_pred`?

tylerjereddy replied:
I line profiled the code and identified the bottleneck, then scoped the part of the algorithm that made sense to make parallel.

With 200,000 features (1,000 records, 50 estimators) on 16 cores the finalization loop has negligible contribution so I didn't want to make code concurrent that didn't need to be:

```
   619         2     688182.1 344091.1     99.8          all_pred = Parallel(n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem")(
   620                                                           delayed(self._get_oob_pred_parallel)(e, n_samples, n_samples_bootstrap, X)
   621         1          0.2      0.2      0.0              for e in self.estimators_)
   622                                           
   623        51         26.0      0.5      0.0          for unsampled_indices, y_pred in all_pred:
   624        50       1111.8     22.2      0.2              oob_pred[unsampled_indices, ...] += y_pred
   625        50        386.2      7.7      0.1              n_oob_pred[unsampled_indices, :] += 1
```

If I bump that up to 5,000 estimators it is even more dramatic:

```
   619         2   69161602.1    3e+07     99.9          all_pred = Parallel(n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem")(
   620                                                           delayed(self._get_oob_pred_parallel)(e, n_samples, n_samples_bootstrap, X)
   621         1          0.2      0.2      0.0              for e in self.estimators_)
   622                                           
   623      5001        803.0      0.2      0.0          for unsampled_indices, y_pred in all_pred:
   624      5000      34326.1      6.9      0.0              oob_pred[unsampled_indices, ...] += y_pred
   625      5000      12399.4      2.5      0.0              n_oob_pred[unsampled_indices, :] += 1
```

adam2392 replied:
An alternative way that's more in-line w/ the existing `predict_proba` parallelization is

1. The following private function:

```
def _accumulate_oob_prediction(predict, tree, X, n_samples, n_samples_bootstrap, out, out_count, random_state, lock):
    """
    This is a utility function for joblib's Parallel.

    It can't go locally in ForestClassifier or ForestRegressor, because joblib
    complains that it cannot pickle it when placed there.
    """
    unsampled_indices = _generate_unsampled_indices(
        random_state,
        n_samples,
        n_samples_bootstrap,
    )

    prediction = predict(tree, X[unsampled_indices, :])
    with lock:
        out[unsampled_indices, ...] += prediction
        out_count[unsampled_indices, :] += 1
```

2. The `_compute_oob_predictions` LOC is replaced with:

```
       n_samples_bootstrap = _get_n_samples_bootstrap(
            n_samples,
            self.max_samples,
        )
        lock = threading.Lock()
        Parallel(n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem")(
            delayed(_accumulate_oob_prediction)(self._get_oob_predictions, e, X, n_samples, n_samples_bootstrap,
                                                oob_pred, n_oob_pred, e.random_state, lock)
            for e in self.estimators_
        )
```

However, I think one of the maintainers will chime in.

tylerjereddy replied:
I didn't see a reason to pass in a data structure and use a lock to accumulate in it. If there's a compelling performance argument I'd be happy to. I did see that infrastructure and decided it wasn't quite what I wanted, but I don't really care too much as long as the performance improvement holds steady.

thomasjpfan replied:
Given that this is using multithreading with `require="sharedmem"`, I think the memory issue comes from having to store all the predictions all at once in `all_preds`.

I suspect that having each worker update `oob_pred` with a lock would resolve the memory issue in https://github.com/scikit-learn/scikit-learn/pull/28442#issuecomment-2334910902

tylerjereddy replied:
I think the lock contention is too slow to scale well for our use case. I committed the diff below the fold locally and repeated the two benchmarking incantations, and the results were not particularly exciting. The memory cost is gone, but so is the scaling with more concurrency.

<details>

```diff
--- a/sklearn/ensemble/_forest.py
+++ b/sklearn/ensemble/_forest.py
@@ -557,14 +557,16 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
             (accuracy score).
         """
 
-    def _get_oob_pred_parallel(self, estimator, n_samples, n_samples_bootstrap, X):
+    def _get_oob_pred_parallel(self, oob_pred, n_oob_pred, estimator, n_samples, n_samples_bootstrap, X, lock):
         unsampled_indices = _generate_unsampled_indices(
             estimator.random_state,
             n_samples,
             n_samples_bootstrap,
         )
         y_pred = self._get_oob_predictions(estimator, X[unsampled_indices, :])
-        return unsampled_indices, y_pred
+        with lock:
+            oob_pred[unsampled_indices, ...] += y_pred
+            n_oob_pred[unsampled_indices, :] += 1
 
     def _compute_oob_predictions(self, X, y):
         """Compute and set the OOB score.
@@ -607,17 +609,12 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
             self.max_samples,
         )
 
-        all_pred = Parallel(
-            n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem"
-        )(
-            delayed(self._get_oob_pred_parallel)(e, n_samples, n_samples_bootstrap, X)
+        lock = threading.Lock()
+        Parallel(n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem")(
+            delayed(self._get_oob_pred_parallel)(oob_pred, n_oob_pred, e, n_samples, n_samples_bootstrap, X, lock)
             for e in self.estimators_
         )
 
-        for unsampled_indices, y_pred in all_pred:
-            oob_pred[unsampled_indices, ...] += y_pred
-            n_oob_pred[unsampled_indices, :] += 1
-
         for k in range(n_outputs):
             if (n_oob_pred == 0).any():
                 warn(
```

</details>

First incantation:

```
| Change   | Before [0b0b90b7] <main>   | After [264d74de] <treddy_issue_28059>   |   Ratio | Benchmark (Parameter)                                                |
|----------|----------------------------|-----------------------------------------|---------|----------------------------------------------------------------------|
| -        | 1.16±0.02s                 | 838±10ms                                |    0.72 | ensemble.RandomForestClassifierBenchmark.time_fit('sparse', 8, True) |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

Second incantation at higher concurrency/larger problems:

```
BENCHMARKS NOT SIGNIFICANTLY CHANGED.
```

I guess I could try thinking about other models of concurrency a bit. I think when I played with other approaches they had other annoying tradeoffs, but maybe I should double check..

tylerjereddy replied:
I dunno, the other backends like `loky` do indeed seem to have serialization issues and so on. We could try to split the array and recombine it at the end--is that worth exploring?

thomasjpfan replied:
Ah I see. Can you try joblib's new generator support? The API is almost the same as what you already have but with `Parallel(..., return_as="generator")`

Reference: https://joblib.readthedocs.io/en/latest/auto_examples/parallel_generator.html

tylerjereddy committed:
MAINT: PR 28442 revisions

* linter cleanups

adam2392 reviewed (commented):
I verified that there was a noticeable improvement on my local computer when running oob (as expected). This seems like a valuable improvement for oob computation for forests.

adam2392 reviewed (commented):
Hi @tylerjereddy can you include a changelog entry in https://github.com/scikit-learn/scikit-learn/blob/main/doc/whats_new/v1.6.rst documenting the speedup?

Otw, I think this is a nice improvement to the oob setting.

adam2392 commented:
Perhaps @glemaitre and @thomasjpfan can take a look as well?

tylerjereddy commented:
Oh right, I know Thomas. I guess if I were to criticize this I might wonder if some of the other tree-based methods might benefit (is the infrastructure general enough?), but maybe that could be done later.

Did we get the benchmarks tested? It looks I didn't actually sort out the `asv` run, but should probably confirm that works.

tylerjereddy commented:
One other thing to confirm might be that the tests properly flush the new parallel code path? Does that happen automatically since `n_jobs` is already used productively elsewhere?

adam2392 commented:
> Oh right, I know Thomas. I guess if I were to criticize this I might wonder if some of the other tree-based methods might benefit (is the infrastructure general enough?), but maybe that could be done later.

I think the only other place I can currently think of is in `_bagging.py`, but that is a separate file and the implementation will slightly differ I imagine. I feel like that can be done later, but perhaps someone else disagrees.

> Did we get the benchmarks tested? It looks I didn't actually sort out the `asv` run, but should probably confirm that works.

Yes that would be nice to run. Would you be able to run the asv benchmark you proposed, or run a similar setting as the one you posted in the [GH issue](https://github.com/scikit-learn/scikit-learn/issues/28059#issue-2064663387), where you also vary the # of samples?

tylerjereddy commented:
For:

`export SKLBENCH_NJOBS=[1,8]`
`asv continuous -E virtualenv -e -b "RandomForestClassifierBenchmark" main treddy_issue_28059`

There is one benchmark that gets faster, `oob` active with concurrency as you might expect, but also some memory cost/overhead associated:

```
| Change   | Before [0b0b90b7] <main>   | After [44d6e409] <treddy_issue_28059>   |   Ratio | Benchmark (Parameter)                                                   |
|----------|----------------------------|-----------------------------------------|---------|-------------------------------------------------------------------------|
| +        | 211M                       | 266M                                    |    1.26 | ensemble.RandomForestClassifierBenchmark.peakmem_fit('dense', 8, True)  |
| +        | 434M                       | 521M                                    |    1.2  | ensemble.RandomForestClassifierBenchmark.peakmem_fit('sparse', 8, True) |
| +        | 210M                       | 246M                                    |    1.17 | ensemble.RandomForestClassifierBenchmark.peakmem_fit('dense', 1, True)  |
| -        | 1.22±0.06s                 | 916±10ms                                |    0.75 | ensemble.RandomForestClassifierBenchmark.time_fit('sparse', 8, True)    |


```

Using the larger scale benchmarks shows a more dramatic performance improvement starting to show up. There is still some memory cost.

`export SKLBENCH_PROFILE=large_scale`
`export SKLBENCH_NJOBS=[16]`
`asv continuous -E virtualenv -e -b "RandomForestClassifierBenchmark" main treddy_issue_28059`

```
| Change   | Before [0b0b90b7] <main>   | After [44d6e409] <treddy_issue_28059>   |   Ratio | Benchmark (Parameter)                                                   |
|----------|----------------------------|-----------------------------------------|---------|-------------------------------------------------------------------------|
| +        | 601M                       | 789M                                    |    1.31 | ensemble.RandomForestClassifierBenchmark.peakmem_fit('dense', 16, True) |
| -        | 4.47±0.08s                 | 2.69±0.07s                              |    0.6  | ensemble.RandomForestClassifierBenchmark.time_fit('sparse', 16, True)   |
```

tylerjereddy commented:
Naive addition of `return_as="generator"` seemed to nullify the performance gains; I tried the fancy new `generator_unordered` as well, but seemed the same. Maybe I need to think about the design a bit more carefully in the generator context though.

<details>

```diff
--- a/sklearn/ensemble/_forest.py
+++ b/sklearn/ensemble/_forest.py
@@ -608,7 +608,7 @@ class BaseForest(MultiOutputMixin, BaseEnsemble, metaclass=ABCMeta):
         )
 
         all_pred = Parallel(
-            n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem", return_as="generator"
+            n_jobs=self.n_jobs, verbose=self.verbose, require="sharedmem", return_as="generator_unordered"
         )(
             delayed(self._get_oob_pred_parallel)(e, n_samples, n_samples_bootstrap, X)
             for e in self.estimators_

```


</details>

tylerjereddy committed:
ENH, MAINT: PR 28442 revisions

* use `return_as="generator_unordered"` to avoid single-core
memory overhead in the parallel random forest OOB implementation

* the above change requires that benchmarks used a newer version
of `joblib`, so adjust accordingly

tylerjereddy committed:
ENH, MAINT: PR 28442 revisions

* To reduce the impact of memory issues with
parallel OOB calculations in Random Forest the
following changes were made:

1) Use a lock to update the prediction-related
data structures, to avoid storing all predictions
in `all_pred` (problematic even when returning
a generator).

2) Split the work into chunks to reduce the sizes
of stored temporary data structures. This also
reduces the amount of lock contention based
on local testing (without chunking, the lock
seems to abrogate any performance improvements
due to concurrent OOB).

* Remove `joblib` version bound change in `asv`
config--no longer needed now that we don't use
fancy generator returns.

tylerjereddy commented on sklearn/ensemble/_forest.py:
```diff
@@ -597,16 +616,19 @@ def _compute_oob_predictions(self, X, y):
-            n_oob_pred[unsampled_indices, :] += 1
+        lock = threading.Lock()
+        sub_ests = []
+        chunk_size = 40
```
@glemaitre @thomasjpfan I've refactored the parallel strategy to reduce the amount of temporary data that gets stored while avoiding severe lock contention by using chunking on top of a lock (locking alone abrogates all performance gains--it seems that giving each thread a "chunk" of work reduces contention vs. whatever `joblib` does by default). I'm not particularly fond of this solution (the magic number here is pretty atrocious), but it does provide clear wins in the `asv` benchmarks that were previously showing tradeoffs on memory consumption.

Repeating the benchmarks from https://github.com/scikit-learn/scikit-learn/pull/28442#issuecomment-2334910902 (on the same machine):

```
export SKLBENCH_NJOBS=[1,8]
time asv continuous -E virtualenv -e -b "RandomForestClassifierBenchmark" main treddy_issue_28059
```

```
| Change   | Before [429d67aa] <main>   | After [2325f154] <treddy_issue_28059>   |   Ratio | Benchmark (Parameter)                                                |
|----------|----------------------------|-----------------------------------------|---------|----------------------------------------------------------------------|
| -        | 1.10±0.04s                 | 910±6ms                                 |    0.83 | ensemble.RandomForestClassifierBenchmark.time_fit('sparse', 8, True) |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

```
export SKLBENCH_PROFILE=large_scale
export SKLBENCH_NJOBS=[16]
time asv continuous -E virtualenv -e -b "RandomForestClassifierBenchmark" main treddy_issue_28059
```

```
| Change   | Before [429d67aa] <main>   | After [2325f154] <treddy_issue_28059>   |   Ratio | Benchmark (Parameter)                                                 |
|----------|----------------------------|-----------------------------------------|---------|-----------------------------------------------------------------------|
| -        | 3.56±0.01s                 | 2.50±0.06s                              |     0.7 | ensemble.RandomForestClassifierBenchmark.time_fit('sparse', 16, True) |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```

I'm not convinced my time was well spent here though--I tried a lot of different things, and at the end of the day it seems like what we really need is a highly-performant "atomic accumulator" data structure where we can quickly accumulate/reduce the predictions without unreasonable lock contention nor usage of a "magic number" for chunking.

Also, the `asv` benchmarks for the project may not necessarily provide a solid proxy for our in-house use case (ratio of features:records around 900_000:900).  If I repeat the informal scaling script from the original ticket on the latest version of this branch, while more impressive from the `asv` standpoint, it looks much less impressive for the original real-world use case proxy:

<details>

![image](https://github.com/user-attachments/assets/cfc21ff2-fe5d-4839-9eb2-c0d188b6c770)


</details>

That might be a little better than before, but we'd likely still have to vendor our own solution for the foreseeable future.

It seems like there are at least two routes forward:

1) Accept that the `asv` benchmarks aren't particularly great proxies for what we're trying to accomplish, and accept the memory costs for the infrastructure changes that perform well in the "real world." You do have a separate non-asv `benchmarks` folder, so this may be a situation you are not entirely unfamiliar with. That said, I would agree that it would be really nice if faster `asv` proxy benchmarks could do the trick. Perhaps with a much larger feature:record ratio they might.
2) Continue to use `asv` as the "gatekeeper"/standard, and accept that for now this is a step toward using parallel OOB, but that what we really need is a solution with both reasonable memory footprint and excellent real-world scaling behavior.

So, it would probably be good to decide what path you'd like to take here.
